### PR TITLE
Bug 1570026 - Monitor action for snippets

### PR DESCRIPTION
**Pull request by k88hudson**
This adds a new user action to ASRouter that allows users to directly authenticate with Firefox Monitor via a URL, similar to how the What's New Page does it.

I have a few questions about the best way to implement this, noted below in comments, before I add tests.

### CODE REVIEW

File: lib/ASRouter.jsm [hunk 1]
```diff
@@ -1767,7 +1767,62 @@ class _ASRouter {
...
+          ),
+          csp: null,
+        });
+        break;
+    }
+  }
+
+  async fetchFlowParams({ utm_term, entrypoint, form_type }) {
+    let result = {};
+    const fxaEndpoint = Services.prefs.getStringPref(
```
Comment (k88hudson) [inline lib/ASRouter.jsm]: Do we want to use this Firefox accounts endpoint or rely on the snippets service to provide one?

File: lib/SnippetsTestMessageProvider.jsm [hunk 2]
```diff
@@ -90,6 +90,25 @@ const MESSAGES = () => [
...
+  {
+    id: "SIMPLE_TEST_MONITOR",
+    template: "simple_snippet",
+    content: {
+      icon: TEST_ICON,
+      icon_dark_theme: TEST_ICON_BW,
+      button_label: "Get monitor",
+      button_action: "ENABLE_FIREFOX_MONITOR",
+      button_action_args: {
+        utm_term: "monitor-snippet-test",
```
Comment (k88hudson) [inline lib/SnippetsTestMessageProvider.jsm]: Not sure what the correct values for any of these should be; should they all just come from the snippets service?

File: lib/ASRouter.jsm [hunk 3]
```diff
@@ -1767,7 +1767,62 @@ class _ASRouter {
...
+          utm_campaign,
+          entrypoint,
+          form_type,
+        } = action.data;
+        const { flowBeginTime, flowId } = await this.fetchFlowParams({
+          entrypoint,
+          utm_term,
+          form_type,
+        });
+        const monitorURL = `https://monitor.firefox.com/oauth/init?utm_source=${utm_source}&utm_campaign=${utm_campaign}&form_type=${form_type}&entrypoint=${entrypoint}&flowBeginTime=${flowBeginTime}&flowId=${flowId}`;
```
Comment (k88hudson) [inline lib/ASRouter.jsm]: I'm curious what you think about passing the whole URL with templates for `flowBeginTime` and `flowId` (which have to come from Firefox) v.s. passing variables:

With templated URL: 
```json
{
  "button_action": "ENABLE_FIREFOX_MONITOR",
  "button_action_args": "https://monitor.firefox.com/oauth/init?utm_source=snippets&utm_campaign=foo&${flowBeginTime}&flowId=${flowId}"
}
```

With variables: 
```json
{
  "button_action": "ENABLE_FIREFOX_MONITOR",
  "button_action_args": {
    "utm_source": "snippets",
    "utm_campaign": "foo"
  }
}
```

I think passing the whole URL could give you more flexibility in changing different parts of the URL remotely, but variables might be less error-prone.

Comment (glogiotatidis) [reply lib/ASRouter.jsm]: I prefer URL for the flexibility. Either way works fine for me. Variables or URL will be handled in the backend and will not be directly editable by Snippet Editors so errors will be limited. Thanks for reaching out! :cake:

[the remark this answers is not in the document]
Comment (piatra) [reply lib/ASRouter.jsm]: Note that the two ways of constructing the string _can_ result in different outputs:
For a `utm_campaign` value of `Firefox Monitor`
* template literals result in `&utm_campaign=Firefox Monitor` which then becomes `&utm_campaign=Firefox%20Monitor`
* URLSearchParams will do `&utm_campaign=Firefox+Monitor`

File: lib/SnippetsTestMessageProvider.jsm (hunk 2 above)
Comment (piatra) [reply lib/SnippetsTestMessageProvider.jsm]: They're all strings so it seems ok.

File: lib/ASRouter.jsm [hunk 4]
```diff
@@ -1767,7 +1767,62 @@ class _ASRouter {
...
         });
         break;
+      case ra.ENABLE_FIREFOX_MONITOR:
+        const {
+          utm_term,
+          utm_source,
+          utm_campaign,
+          entrypoint,
+          form_type,
+        } = action.data;
```
Comment (piatra) [inline lib/ASRouter.jsm]: This should be `action.data.args`.

File: lib/ASRouter.jsm [hunk 5]
```diff
@@ -1767,7 +1767,62 @@ class _ASRouter {
...
+    const fxaEndpoint = Services.prefs.getStringPref(
+      "browser.newtabpage.activity-stream.fxaccounts.endpoint",
+      "https://accounts.firefox.com"
+    );
+    try {
+      const url = new URL(
+        `${fxaEndpoint}/metrics-flow?entrypoint=${entrypoint}&form_type=${form_type}&utm_term=${utm_term}`
+      );
+      const response = await fetch(url, { credentials: "omit" });
+      if (response.status === 200) {
```
[the remark this answers is not in the document]
Comment (k88hudson) [reply lib/ASRouter.jsm]: Good question, this is what we're doing for the what's new page

File: lib/ASRouter.jsm (hunk 1 above)
Comment (shane-tomlinson) [reply lib/ASRouter.jsm]: Is this targeting Firefox 70? Over in https://bugzilla.mozilla.org/show_bug.cgi?id=1568949 and https://bugzilla.mozilla.org/show_bug.cgi?id=1556276, there is talk of having newtab automatically respect the `identity.fxaccounts.remote.root` values, it seems like this should use that same code, which I imagine is bbased off of `identity.fxaccounts.remote.root`?

File: lib/ASRouter.jsm (hunk 5 above)
[the remark this answers is not in the document]
Comment (shane-tomlinson) [reply lib/ASRouter.jsm]: /metrics-flow isn't cached and will generate a new response every time, 304 isn't possible, checking for a 200 should be sufficient.

File: lib/ASRouter.jsm (hunk 1 above)
Comment (piatra) [reply lib/ASRouter.jsm]: This will be uplifted to 69.

File: lib/SnippetsTestMessageProvider.jsm [hunk 6]
```diff
@@ -402,6 +415,27 @@ const MESSAGES = () => [
...
+    template: "simple_below_search_snippet",
+    content: {
+      icon: TEST_ICON,
+      title: "See if you've been part of an online data breach.",
+      text: "Firefox Monitor tells you what hackers already know about you.",
+      button_label: "Get monitor",
+      button_action: "ENABLE_FIREFOX_MONITOR",
+      button_action_args: {
+        url:
+          "https://monitor.firefox.com/oauth/init?utm_source=snippets&utm_campaign=monitor-snippet-test&form_type=email&entrypoint=newtab",
```
Comment (glogiotatidis) [inline lib/SnippetsTestMessageProvider.jsm]: I see both url and flowRequestParams. Can you drop a line about when to use what? @k88hudson

Comment (k88hudson) [reply lib/SnippetsTestMessageProvider.jsm]: Yep, I will add some docs. Essentially the URL is everything snippets would be able to provide, the `flowParams` is a few properties needed to pass to a Firefox metrics endpoint (defined via a pref) in order to get the right `flowId`, `flowBeginTime` and `deviceId` back.

**Pull request merged**